File: golang/board/board.go
package board

import (
	"errors"
	"math/rand"
	"strconv"
	"time"
)

const CellDead = 0
const CellAlive = 1

type board struct {
	state [][]int
	rows int
	columns int
}

/* Creates a new Board with the given dimensions. The dimensions, rows and columns,
must be positive integers greater than 0.
Returns a board populated with a random state. */
func NewRandomBoard(rows, columns int) (board, error) {

	if rows < 1 || columns < 1 {
		return board{}, errors.New("rows and columns must be a positive integer greater than 0")
	}

	initState := make([][]int, rows)
	for i := range initState {
		initState[i] = make([]int, columns)
	}

	rand.Seed(time.Now().UnixNano())

	// Populate random state
	for i := range initState {
		for j := range initState[i] {
			initState[i][j] = rand.Intn((1 -0 + 1) + 0)
		}
	}

	return board{state: initState, rows:rows, columns:columns}, nil
}

func NewBoard(initialState [][]int) (board, error) {

	if initialState == nil {
		return board{}, errors.New("initialState cannot be nil")
	}

	if len(initialState) < 1 || len(initialState[0]) < 1 {
		return board{}, errors.New("initialState must contain at least 1 row and 1 column")
	}

	colSize := len(initialState[0])
	for i := 0; i < len(initialState); i++ {
		if colSize != len(initialState[i]) {
			return board{}, errors.New("initialState is a jagged 2D array, initialState cannot be jagged")
		}
		for j := 0; j < len(initialState[i]); j++ {
			cellValue := initialState[i][j]
			if cellValue < 0 || cellValue > 1 {
				return board{}, errors.New("initialState may only contain values 0 or 1")
			}
		}
	}

	return board{state:initialState, rows: len(initialState), columns: len(initialState[0])}, nil
}

func (b *board) Evolve() {
	newState := make([][]int, b.rows)
	for i := range newState {
		newState[i] = make([]int, b.columns)
		for j := range newState[i] {
			newState[i][j] = nextStateForCell(b,i,j)
		}
	}
	b.state = newState
}

func (b *board) State() [][]int {
	return b.state
}

func (b *board) Rows() int {
	return b.rows
}

func (b *board) Columns() int {
	return b.columns
}

func (b *board) PrettyPrint() {
	for i := range b.state {
		for j := range b.state[i] {
			print(" " + strconv.Itoa(b.state[i][j]) + "")
		}
		println()
	}
}

func nextStateForCell(b *board, i,j int) int {

	neighborsAlive := 0
	cellValue := b.state[i][j]
	for x := -1; x <= 1; x++ {
		for y := -1; y <= 1; y++ {
			if i + x < 0 || i + x > (b.rows - 1) || y + j < 0 || y + j > (b.columns - 1) {
				continue
			}
			neighborsAlive += b.state[i + x][y + j]
		}
	}
	neighborsAlive -= cellValue

	if cellValue == CellDead && neighborsAlive == 3 {
		return CellAlive
	} else if cellValue == CellAlive && (neighborsAlive < 2 || neighborsAlive > 3) {
		return CellDead
	} else {
		return cellValue
	}
}

File: ruby/board.rb
class Board

  attr_reader :height, :state, :width

  # Creates a new board with the given width and height with a random
  # initial state
  #
  # Params
  # - height: height or number of rows for the board, must be greater than 0
  # - width: width or number of columns for the board, must be greater than 0
  #
  # Returns a new board
  def self.random(width, height)

    if width.nil? || width < 0
      raise ArgumentError, 'width must be a positive integer > 0'
    end

    if height.nil? || height < 0
      raise ArgumentError, 'height must be a positive integer > 0'
    end

    state = Array.new(height) {Array.new(width)}

    (0...height).each do |row|
      (0...width).each do |cell|
        state[row][cell] = Random.rand(2)
      end
    end
    Board.new(state)
  end

  # Initializes the Board with a provided initial state. The provided initial
  # state should be a 2D array with a minimum height (row) of 1, and width (column)
  # of 1. The 2D array must also not be jagged. The board uses 0 to represent a dead
  # cell and 1 to represent a live cell. Those are the only valid values that can be
  # provided in the initial state. If the initial state is not valid an ArgumentException
  # will be raised.
  #
  # Params:
  # - initial_state: the initial state of the board, cannot be nil
  #
  # Returns: new Board with the provided initial state
  def initialize(initial_state)

    # Validates the initial state is valid, if not an error will be raised
    validate_state(initial_state)

    # If we reach this point the initial state is valid, init instance variables
    @state = initial_state
    @width = initial_state.length
    @height = initial_state[0].length
  end

  # Evolves the board to the next state
  def evolve

    new_state = Array.new(height) {Array.new(width)}

    (0...@height).each do |row|
      (0...@width).each do |cell|
        neighbors_alive = calculate_living_neighbors(row, cell)
        cell_value = @state[row][cell]

        if cell_value == 1 and (neighbors_alive < 2 or neighbors_alive > 3)
          new_state[row][cell] = 0
        elsif cell_value == 0 and neighbors_alive == 3
          new_state[row][cell] = 1
        else
          new_state[row][cell] = cell_value
        end
      end
    end
    @state = new_state
  end

  private
  def validate_state(state)

    if state.nil?
      raise ArgumentError, 'initial_state cannot be nil'
    end
  end

  def calculate_living_neighbors(i, j)

    live_count = 0
    (-1...2).each do |x|
      (-1...2).each do |y|
        if i + x < 0 or i + x > (@height - 1) or y + j < 0 or y + j > (@width - 1)
          # do nothing
        else
          live_count += @state[i + x][y + j]
        end
      end
    end
    live_count -= @state[i][j]
  end
end